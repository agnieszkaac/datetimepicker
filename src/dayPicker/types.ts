import React from "react";

export interface DayProps {
  value: string;
  selected: boolean;
  onPick: (e: React.MouseEvent<HTMLButtonElement>) => void;
}

export interface DayPickerProps {
  value: string;
  monthLength: number;
  onPick: (e: React.MouseEvent<HTMLButtonElement>) => void;
}
