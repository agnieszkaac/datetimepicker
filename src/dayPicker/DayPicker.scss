$size: 25px;

.day-picker {
  display: grid;
  grid-template-columns: repeat(7, $size);
  grid-gap: 3px;
}

.day {
  display: flex;
  justify-content: center;
  align-items: center;
  height: $size;
  width: $size;
  background-color: white;
  color: black;
  border: none;
  outline: none;

  &:hover {
    background-color: silver;
  }
  &.selected {
    background-color: aquamarine;
  }
  &.is-today {
    border: 1px solid blue;
  }
}

