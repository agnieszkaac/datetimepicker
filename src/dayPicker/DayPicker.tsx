import React from "react";
import moment from "moment";

import { DayPickerProps, Day } from "./";

export const DayPicker: React.FunctionComponent<DayPickerProps> = ({
  date,
  viewDate,
  onPick,
}) => {
  return (
    <div className="day-picker">
      {Array.from({ length: moment(viewDate).daysInMonth() }, (_e, i) => (
        <Day
          key={i + 1}
          date={moment(viewDate).date(i + 1)}
          selected={date?.date() === i + 1 && date?.isSame(viewDate, "month")}
          onPick={onPick}
        />
      ))}
    </div>
  );
};
