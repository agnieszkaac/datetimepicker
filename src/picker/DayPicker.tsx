import React from "react";
import cx from "classnames";
import moment from "moment";

import { DayPickerProps, DayProps } from "../types";

const Day: React.FunctionComponent<DayProps> = ({
  value,
  selected,
  isToday,
  onPick,
}) => (
  <button
    className={cx("day", { selected: selected, "is-today": isToday })}
    onClick={onPick}
    value={value}
  >
    {value}
  </button>
);

export const DayPicker: React.FunctionComponent<DayPickerProps> = ({
  value,
  monthLength,
  onPick,
}) => (
  <div className="day-picker">
    {Array.from({ length: monthLength }, (e, i) => (
      <Day
        key={i + 1}
        value={(i + 1).toString()}
        selected={value === (i + 1).toString()}
        isToday={moment().date().toString() === value}
        onPick={onPick}
      />
    ))}
  </div>
);
