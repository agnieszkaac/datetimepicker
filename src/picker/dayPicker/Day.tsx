import React from "react";
import { Moment } from "moment";
import cx from "classnames";

export type DayProps = {
  viewDate: Moment;
  className?: string;
  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Day: React.FunctionComponent<DayProps> = ({
  viewDate,
  className,
  onClick,
}) => (
  <button
    type="button"
    data-testid="day"
    value={viewDate.toISOString()}
    className={cx("day", className)}
    onClick={onClick}
  >
    {viewDate.date()}
  </button>
);
