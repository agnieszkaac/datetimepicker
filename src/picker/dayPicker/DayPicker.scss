@import "../../common";

$size: $size-m;

.day {
  height: $size;
  width: $size;
  color: $text;
  background-color: $background;
  border: none;
  outline: none;
  border-radius: 2px;
  cursor: pointer;

  &:hover {
    background-color: $background-hover;
  }
}

.day-picker {
  @include grid($size, 7);
  & .day {
    &.prev-month,
    &.next-month {
      color: $text-light;
    }
    &.is-today {
      border: 1px solid $border-today;
    }
    &.selected {
      color: $text-selected;
      background-color: $background-selected;
    }
  }
}
.week-days {
  @include grid($size, 7);
  .day-of-week {
    margin: $spacing-m 0;
    text-align: center;
    font-size: $font-size-small;
    line-height: $size-xs;
    font-weight: 600;
    vertical-align: middle;
  }
}
