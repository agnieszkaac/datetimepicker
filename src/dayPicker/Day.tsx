import React from "react";
import moment from "moment";
import cx from "classnames";

import { DayProps } from "./";
import "./DayPicker.scss";

export const Day: React.FunctionComponent<DayProps> = ({
  date,
  selected,
  onPick,
}) => {
  const isToday = date.isSame(moment(), "day");
  return (
    <button
      type="button"
      className={cx("day", { selected: selected, "is-today": isToday })}
      value={date.toISOString()}
      onClick={onPick}
    >
      {date.date()}
    </button>
  );
};
