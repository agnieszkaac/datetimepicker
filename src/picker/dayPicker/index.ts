import { Day } from "./Day";
import DayPicker from "./DayPicker";
import { DayProps, DayPickerProps } from "./types";

export { Day, DayPicker, DayProps, DayPickerProps };
