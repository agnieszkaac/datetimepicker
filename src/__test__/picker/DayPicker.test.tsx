import React from "react";
import MockDate from "mockdate";
import moment, { Moment } from "moment";
import { render } from "@testing-library/react";

import { DayPicker, DayPickerProps } from "../../picker/dayPicker";

describe("<DayPicker /> should", () => {
  let date: Moment | undefined;
  let viewDate: Moment;
  const onPick = jest.fn((_e) => {});
  let props: DayPickerProps;

  beforeEach(() => {
    MockDate.set("2020-01-01");

    date = undefined;
    viewDate = moment();
    props = {
      date,
      viewDate,
      onPick,
    };
  });

  afterEach(() => {
    MockDate.reset();
  });

  it("match snapshot", () => {
    const { asFragment } = render(<DayPicker {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("render 42 days in a grid", () => {
    const { getAllByTestId } = render(<DayPicker {...props} />);
    const days = getAllByTestId("day");

    expect(days.length).toBe(42);
  });

  it("mark day as selected when date prop is passed", () => {
    const { getByText, rerender } = render(<DayPicker {...props} />);
    const day = getByText("15");

    expect(day).not.toHaveClass("selected");

    props.date = moment("2020-01-15");
    rerender(<DayPicker {...props} />);

    expect(day).toHaveClass("selected");
  });

  it("not mark any selected days when selected is not within viewDate", () => {
    props.date = moment();
    props.viewDate = moment("2020-10-10");

    const { asFragment } = render(<DayPicker {...props} />);
    const days = asFragment().querySelectorAll(".selected");

    expect(days.length).toEqual(0);
  });

  it("mark one day as current day when within viewDate", () => {
    const { asFragment } = render(<DayPicker {...props} />);
    const days = asFragment().querySelectorAll(".is-today");

    expect(days.length).toEqual(1);
  });

  it("not mark any current days when today is not within viewDate", () => {
    props.viewDate = moment("2020-10-10");

    const { asFragment } = render(<DayPicker {...props} />);
    const days = asFragment().querySelectorAll(".is-today");

    expect(days.length).toEqual(0);
  });

  it("render 3 prev-month and 8 next-month days", () => {
    const { container } = render(<DayPicker {...props} />);
    const prevDays = container.querySelectorAll(".prev-month");
    const nextDays = container.querySelectorAll(".next-month");

    expect(prevDays.length).toEqual(3);
    expect(nextDays.length).toEqual(8);
  });
});
