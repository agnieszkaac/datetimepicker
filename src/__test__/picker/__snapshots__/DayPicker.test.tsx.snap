// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DayPicker /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="week-days"
  >
    <span
      class="week-day"
      data-testid="week-day"
    >
      Sun
    </span>
    <span
      class="week-day"
      data-testid="week-day"
    >
      Mon
    </span>
    <span
      class="week-day"
      data-testid="week-day"
    >
      Tue
    </span>
    <span
      class="week-day"
      data-testid="week-day"
    >
      Wed
    </span>
    <span
      class="week-day"
      data-testid="week-day"
    >
      Thu
    </span>
    <span
      class="week-day"
      data-testid="week-day"
    >
      Fri
    </span>
    <span
      class="week-day"
      data-testid="week-day"
    >
      Sat
    </span>
  </div>
  <div
    class="day-picker"
    data-testid="day-picker"
  >
    <button
      class="day prev-month"
      data-testid="day"
      type="button"
      value="2019-12-29T00:00:00.000Z"
    >
      29
    </button>
    <button
      class="day prev-month"
      data-testid="day"
      type="button"
      value="2019-12-30T00:00:00.000Z"
    >
      30
    </button>
    <button
      class="day prev-month"
      data-testid="day"
      type="button"
      value="2019-12-31T00:00:00.000Z"
    >
      31
    </button>
    <button
      class="day is-today"
      data-testid="day"
      type="button"
      value="2020-01-01T00:00:00.000Z"
    >
      1
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-02T00:00:00.000Z"
    >
      2
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-03T00:00:00.000Z"
    >
      3
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-04T00:00:00.000Z"
    >
      4
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-05T00:00:00.000Z"
    >
      5
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-06T00:00:00.000Z"
    >
      6
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-07T00:00:00.000Z"
    >
      7
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-08T00:00:00.000Z"
    >
      8
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-09T00:00:00.000Z"
    >
      9
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-10T00:00:00.000Z"
    >
      10
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-11T00:00:00.000Z"
    >
      11
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-12T00:00:00.000Z"
    >
      12
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-13T00:00:00.000Z"
    >
      13
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-14T00:00:00.000Z"
    >
      14
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-15T00:00:00.000Z"
    >
      15
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-16T00:00:00.000Z"
    >
      16
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-17T00:00:00.000Z"
    >
      17
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-18T00:00:00.000Z"
    >
      18
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-19T00:00:00.000Z"
    >
      19
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-20T00:00:00.000Z"
    >
      20
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-21T00:00:00.000Z"
    >
      21
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-22T00:00:00.000Z"
    >
      22
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-23T00:00:00.000Z"
    >
      23
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-24T00:00:00.000Z"
    >
      24
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-25T00:00:00.000Z"
    >
      25
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-26T00:00:00.000Z"
    >
      26
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-27T00:00:00.000Z"
    >
      27
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-28T00:00:00.000Z"
    >
      28
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-29T00:00:00.000Z"
    >
      29
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-30T00:00:00.000Z"
    >
      30
    </button>
    <button
      class="day"
      data-testid="day"
      type="button"
      value="2020-01-31T00:00:00.000Z"
    >
      31
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-01T00:00:00.000Z"
    >
      1
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-02T00:00:00.000Z"
    >
      2
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-03T00:00:00.000Z"
    >
      3
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-04T00:00:00.000Z"
    >
      4
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-05T00:00:00.000Z"
    >
      5
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-06T00:00:00.000Z"
    >
      6
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-07T00:00:00.000Z"
    >
      7
    </button>
    <button
      class="day next-month"
      data-testid="day"
      type="button"
      value="2020-02-08T00:00:00.000Z"
    >
      8
    </button>
  </div>
</DocumentFragment>
`;
